Styling alternatives (#187)
aws-amplify/amplify-ui: docs/src/pages/ui/theme/alternativeStyling/index.page.mdx

---
title: Alternative Styling
---

import Block from 'amplify-docs/src/components/Block';
import BlockSwitcher from 'amplify-docs/src/components/BlockSwitcher';

One of the goals of Amplify is to be a good citizen of the environment it exists in. As such Amplify works hard to not interfere with other tools that are being used.

## CSS In JS

> There are many CSS In JS libraries out there that give a number of benefits. Amplify does not use a CSS In JS library to implement its theming system but operates well with the major CSS In JS libraries out there. Below are a few of the CSS In JS libraries out there and a description of how they interact with Amplify's theming system.

Reviewer: 'CSS in JS libraries out there' twice in succession, we can probably change the second sentence to just say 'Below are a few libraries with descriptions of how they work with Amplify's theming system.'

#### Usage

```jsx
import React from 'react';
import styled from 'styled-components';
import { View } from '@aws-amplify/ui-react';

const StyledView = styled(View)`
  background-color: black;
  font-size: 32px;
`;

//use it like any other amplify-ui component
<StyledView color="red" className="my-styled-view">
  This is my Styled View
</StyledView>;
```

#### Interactions

There are various ways to customize an Amplify component and Styled Components will interact with these customizations in the following ways:

##### Amplify Styling Props

These are styling props that can be used directly on an Amplify component which will affect a single style property. Examples of styling props are `color` or `fontWeight`. These Amplify styling props will take precedence over Styled Component styling.

```jsx
import React from 'react';
import styled from 'styled-components';
import { View } from '@aws-amplify/ui-react';

const StyledView = styled(View)`
  color: blue;
`;

<StyledView color="red">Using Styling props</StyledView>;
```

> In the example above, the color of the view will be set to `red` because of the Amplify styling prop. [Here]() is a list of all Amplify styling props.

Reviewer: TODO for me: create that page that 'Here' links to

##### Amplify variation props

> These are Amplify props that are added to certain Amplify components to change the look and/or behavior of the Amplify component. Examples of these variation props are `size` or `variation`.

Reviewer: We could make this a bit tighter:

'These props change the look and/or behavior of certain Amplify components. Examples include `size` and `variation`.'

```jsx
import React from 'react';
import styled from "styled-components";
import { Button } from '@aws-amplify/ui-react';

const StyledButton = styled(Button)`
  color: blue
`;

<StyledButton variation="primary'>Primary Button</StyledButton>
```

> In the example above, the color of the button will be set to `white`, because the primary variation of the button has a color set to `white` and it takes precedence over the Styled Components styling. A list of Amplify variations can be found on [here]().

Reviewer: Interesting, is this always the case or is it just because with variations our selector has a higher precedence? If you used a button without a variation would the color be blue?

Author: Yes this is because our variation styling rules have a higher precedence, without the variation on the button the library styling would take precedence over ours and the button would be color blue.

Reviewer 2: Couldn't the customer use a data attribute selector to make it work, like this?
```
const StyledButton = styled(Button)`
  &[data-variation='primary']: {
    color: blue
   }
`;
```

Reviewer 2: This makes me wonder if we should try to have the same specificity for all of our custom rules... The attribute selector rules certainly complicate things.

Reviewer: Yea I'm thinking that as well. I guess that means we have to go with BEM or BEM-like syntax then?

Author: yes the developer can use data attribute selectors to make their rules take precedence. In this guide I was trying to describe the experience that they would run into if they simply use a CSS in JS library without putting in extra effort to work well with amplify.  It seems like there are variations and work arounds for anything they might want to do, but I was trying to not branch off into too many tangents for this guide.

##### Amplify Custom ClassNames

These are custom classnames added onto Amplify components and can be used with CSS styling rules to modify the look of an Amplify component.

```jsx
import React from 'react';
import styled from 'styled-components';
import { View } from '@aws-amplify/ui-react';

const styledView = styled(View)`
  color: blue;
`;

<styledView className="my-styled-view">My Styled View</styledView>;
```

```CSS
/* External Style Sheet */
.my-styled-view {
  color: red;
}
```

> In the example above, the color of the view will be `blue` because the stlyed component will take precedence over the classname CSS.

Reviewer 2: ```suggestion
In the example above, the color of the view will be `blue` because the styled component will take precedence over the classname CSS.
```

Reviewer 2: Can you add an explanation of why the styled component CSS will take precedence over the className CSS?

##### CSS Root Variable Overrides

> Amplify theming provides many CSS variables to customize the look and feel of the entire application. These CSS variables can be used at a root level to override all instances of component styling.

Reviewer 2: ```suggestion
Amplify theming provides many CSS variables to customize the look and feel of the entire application. These CSS variables can be used with the `:root` CSS psuedo-class to override all instances of component styling.
```

```jsx
import React from 'react';
import styled from 'styled-components';
import { Button } from '@aws-amplify/ui-react';

const StyledButton = styled(Button)`
  color: blue;
`;

<StyledButton>Button</StyledButton>;
```

```CSS
/* External Style Sheet */
:root {
  --amplify-components-button-color: red;
}
```

> In the example above, the button color will be set to `blue` because the Styled Component will take precedence over the Amplify CSS variable.

Reviewer: Similar question to above, is this true for all CSS variable overrides?

Author: This is true because of styling precedence, but unfortunately styling precedence isn't completely consistent because if they target the variables that are used in rules with a higher precedence then the variables will create changes that override the css in js changes. Some of these variables would be for variations or states (i.e. hover). I'm not sure what the best way to express this to the user would be.

</Block>

<Block name="JSS">

> [JSS](https://github.com/cssinjs/jss/tree/master/packages/react-jss) is a lib for generating Style Sheets with Javascript.

Reviewer: nit: we should probably be consistent with 'library' and 'lib'

These are styling props that can be used directly on an Amplify component which will affect a single style property. Examples of styling props are `color` or `fontWeight`. These Amplify styling props will take precedence over JSS styling.

```jsx
import React from 'react';
import { createUseStyles } from 'react-jss';
import { View } from '@aws-amplify/ui-react';

const useStyles = createUseStyles({
  styledText: {
    color: 'blue',
  },
});

const classes = useStyles();

<Text className={classes.styledText} color="red">
  This is my JSS Text
</Text>;
```

In the example above, the color of the view will be set to `red` because of the Amplify styling prop. [Here]() is a list of all Amplify styling props.

##### Amplify variation props

These props change the look and/or behavior of certain Amplify components. Examples include `size` and `variation`.

```jsx
import React from 'react';
import { createUseStyles } from 'react-jss';
import { Button } from '@aws-amplify/ui-react';

const useStyles = createUseStyles({
  styledButton: {
    color: 'blue',
  },
});

const classes = useStyles();

<Button className={classes.styledButton} variation="primary">
  Primary Button
</Button>;
```

> In the example above, the color of the button will be set to `white`, because the primary variation of the button has a color set to `white` and it takes precedence over the JSS styling. A list of Amplify variations can be found on [here]().

Reviewer 2: Does JSS support `&[data-attribute]` selectors?

Author: yes `&[data-attribute]` selectors are supported. In this guide I was aiming to just describe the default behavior and the experience I would expect developers would run into if they just use CSS in JS libraries without specifically trying to interact with amplify's styling

  styledText: {
    color: 'blue',
  },
});

const classes = useStyles();
const classNames = `${classes.styledText} my-styled-text`;

<Text className={classNames}>This is my JSS Text </Text>;
```

```CSS
/* External Style Sheet */
.my-styled-text {
  color: red;
}
```

In the example above, the color of the text will be `blue` because the JSS styling will take precedence over the classname CSS.

##### CSS Root Variable Overrides

Amplify theming provides many CSS variables to customize the look and feel of the entire application. These CSS variables can be used at a root level to override all instances of component styling.

```jsx
import React from 'react';
import { createUseStyles } from 'react-jss';
import { Button } from '@aws-amplify/ui-react';

const useStyles = createUseStyles({
  styledButton: {
    color: 'blue',
  },
});

const classes = useStyles();

<Button className={classes.styledButton}>Button</Button>;
```

```CSS
/* External Style Sheet */
:root {
  --amplify-components-button-color: red;
}
```

In the example above, the button color will be set to `blue` because the JSS styling will take precedence over the Amplify CSS variable.

</Block>

<Block name="Emotion">

> [Emotion](https://github.com/emotion-js/emotion) is a performant and flexible CSS-in-JS library. It allows you to style apps quickly with string or object styles.

Reviewer: We might want to be a bit more objective in these descriptions so it doesn't look like we are endorsing one library over another.

Author: Yeah probably, for the most part I just took the first sentance or two of their description in their readme file and that is what I used for the description. I'll rework this one to be more objective.

#### Usage

```jsx
/** @jsxImportSource @emotion/react */
import { jsx, css } from '@emotion/react';
import { Text } from '@aws-amplify/ui-react';

const styles = {
  color: 'blue',
};

//use the Emotion css function and css prop
<Text css={css(styles)}>This is my Emotion Text </Text>;
```

#### Interactions

There are various ways to customize an Amplify component and Emotion will interact with these customizations in the following ways:

##### Amplify Styling Props

These are styling props that can be used directly on an Amplify component which will affect a single style property. Examples of styling props are `color` or `fontWeight`. These Amplify styling props will take precedence over Emotion styling.

```jsx
/** @jsxImportSource @emotion/react */
import { jsx, css } from '@emotion/react';
import { Text } from '@aws-amplify/ui-react';

const styles = {
  color: 'blue',
};

> <Text css={css(styles)} color="red">

Reviewer: The `css` prop works?? Cool. 

![cool](https://media.giphy.com/media/62PP2yEIAZF6g/giphy.gif)

>   This is my Emotion Text{' '}

Reviewer: Is `{' '}` something special here?

Author: nope, pulling that out

[Aphrodite](https://github.com/Khan/aphrodite) is a Framework-agnostic CSS-in-JS with support for server-side rendering, browser prefixing, and minimum CSS generation.

#### Usage

```jsx
import { StyleSheet, css } from 'aphrodite';
import { Text } from '@aws-amplify/ui-react';

const styles = StyleSheet.create({
  blue: {
    color: 'blue',
  },
});

//use the Aphrodite css function and the className prop
<Text className={css(styles.blue)}>This is an aphrodite Text component</Text>;
```

#### Interactions

There are various ways to customize an Amplify component and Aphrodite will interact with these customizations in the following ways:

##### Amplify Styling Props

These are styling props that can be used directly on an Amplify component which will affect a single style property. Examples of styling props are `color` or `fontWeight`. Aphrodite will take precedence over these Amplify styling props.

```jsx
import { StyleSheet, css } from 'aphrodite';
import { Text } from '@aws-amplify/ui-react';

const styles = StyleSheet.create({
  blue: {
    color: 'blue',
  },
});

<Text className={css(styles.blue)} color="red">
  This is an aphrodite Text component
</Text>;
```

> In the example above, the color of the view will be set to `blue` because of the Aphrodite styling will take precedence. [Here]() is a list of all Amplify styling props.

Reviewer: It looks like this is the only library where this use-case will be 'blue' instead of 'red', is that correct?

Author: yes by default aphrodite adds !important to all of their styling, there is a config option to turn this off but I just have the default behavior covered here.

Reviewer: ohhh interesting.... should we make a note of this in the docs?

Author: yes we should probably make a note of this, what I didn't want to do is write two docs for Aphrodite where one is how this behaves by default and the other is how this behaves if you take off !important.  Should I just make a note that this is happening because of !important or should I also note that Aphrodite can be configured to take !important off in which case it behaves more like the others?

Reviewer: Yea I think just make a note about the behavior and how they can change that with the aphrodite config.

##### Amplify variation props

These are Amplify props that are added to certain Amplify components to change the look and/or behavior of the Amplify component. Examples of these variation props are `size` or `variation`.

```jsx
import { StyleSheet, css } from 'aphrodite';
import { Button } from '@aws-amplify/ui-react';

const styles = StyleSheet.create({
  blue: {
    color: 'blue',
  },
});

<Button className={css(styles.blue)} variation="primary">
  Primary Button
</Button>;
```

> In the example above, the color of the button will be set to `blue`, because the Aphrodite styling will take precedence over the Amplify variation styling. A list of Amplify variations can be found on [here]().

Reviewer: Same here, all other libraries this would be 'white', is that expected?

Author: yes by default aphrodite adds !important to all of their styling, there is a config option to turn this off but I just have the default behavior covered here.
